docs/source/tutorials/tutorial_tensordot_with_tc.rst

```diff
@@ -0,0 +1,117 @@
+see how to tune the CUDA code to a better performance. All of this is possible with
+only 3-4 lines of code. Let's get started.
+
+For this tutorial, you will need to install Tensor Comprehensions binary. You can
+get binary builds of Tensor Comprehensions with ``conda install -y -c pytorch -c prigoyal tensor_comprehensions``.
+
+About TensorDot operation
+-------------------------
+
+First, we find an operation that we want to generate fast CUDA code for. A lot of
+operations like convolution, pooling are standard and have CUDA code easily available, so
+rather we are going to pick a new and different operation. How do we find a new operation?
+
+**Sources**: Maybe there is a research paper idea you have like KRU or there is a
+numpy operation that is interesting to you and is needed in Machine Learning model.
+As per Numpy docs on linear algebra, tensordot seems like an interesting operation
+`TensorDot <https://docs.scipy.org/doc/numpy/reference/generated/numpy.tensordot.html#numpy.tensordot>`_.
+
+**The TensorDot operation**
+
+Assume that we have two tensors, one with dimension :code:`(N, C1, C2, H, W)` and one with dimension
+:code:`(N, C2, C3, H, W)`, and we want to do a gemm-type computation on the :code:`C`
+dimensions to get an output of shape :code:`(N, C1, C3, H, W)`. Basically, for each
+:code:`(N, H, W)` combination, we want to do a reduction from :code:`(C1, C2) * (C2, C3) = (C1, C3)`.
+
+So basically, this operation can be represented as `N x H x W` independent gemms and one could try to
+write batched gemm kernel for it. But does that guarantee good performance? What if the
+tensor sizes are like this: :code:`N=32, C1=512, C2=8, C3=2, H=28, W=28` i.e.
+the value of :code:`C1` is pretty large compared to :code:`C2` / :code:`C3`.
+
+Let's see how we can get the CUDA kernel for such operation and then tune the kernel.
+
+Step 1: Write TC for TensorDot Operation
+----------------------------------------
+
+First step is to express the Tensordot operation in TC language. For more information on how to do
+so, you can refer to our `Documentation <https://facebookresearch.github.io/TensorComprehensions/index.html>`_
+and also find various TC examples `here <https://facebookresearch.github.io/TensorComprehensions/framework/pytorch_integration/layers_database.html>`_.
+
+.. code-block:: python
+
+    # import tc and torch both
+    import tensor_comprehensions as tc
+    import torch
+    # define the operation as TC language
+    lang = """
+    def tensordot(float(N, C1, C2, H, W) I0, float(N, C2, C3, H, W) I1) -> (O) {
+        O(n, c1, c3, h, w) +=! I0(n, c1, c2, h, w) * I1(n, c2, c3, h, w)
+    }
+    """
+
+Step 2: Register operation with TC
+----------------------------------
+
+Now, we will use the TC string and register it with the TC backend by calling :code:`tc.define`.
+
+.. code-block:: python
+
+    # register the lang with TC backend
+    tensordot = tc.define(lang, name="tensordot")
```

Comment:
Mention that names should match?

Reply:
there is a follow up note for this: see the below figure

<img width="733" alt="screen shot 2018-02-27 at 2 07 45 pm" src="https://user-images.githubusercontent.com/13488275/36749053-a116946e-1bc7-11e8-868f-956f340e36dd.png">